boot/seal.go

```diff
@@ -551,13 +559,46 @@ func resealFallbackObjectKeys(pbc predictableBootChains, authKeyFile string, rol
 	return nil
 }
 
+// recoveryModesForSystems returns a map for recovery modes for recovery systems
+// mentioned in the modeenv. The returned map contains both tested and candidate
+// recovery systems
+func modesForSystems(modeenv *Modeenv) map[string][]string {
+	if len(modeenv.GoodRecoverySystems) == 0 && len(modeenv.CurrentRecoverySystems) == 0 {
+		return nil
+	}
+
+	systemToModes := map[string][]string{}
+
+	// first go through tested recovery systems
+	modesForTestedSystem := []string{ModeRecover, ModeFactoryReset}
+	// tried systems can only boot to recovery mode
+	modesForCandidateSystem := []string{ModeRecover}
+
+	// go through recovery systems that were tested
+	for _, sys := range modeenv.GoodRecoverySystems {
+		systemToModes[sys] = modesForTestedSystem
+	}
+	// go through current recovery systems which can contain systems that
+	// are only being tried
+	for _, sys := range modeenv.CurrentRecoverySystems {
+		if _, ok := systemToModes[sys]; !ok {
+			// since we already went through tested system, this is
+			// a candidate system
+			systemToModes[sys] = modesForCandidateSystem
+		}
+	}
+	return systemToModes
+}
+
 // TODO:UC20: this needs to take more than one model to accommodate the remodel
 // scenario
-func recoveryBootChainsForSystems(systems []string, trbl bootloader.TrustedAssetsBootloader, modeenv *Modeenv, includeTryModel bool) (chains []bootChain, err error) {
-
+func recoveryBootChainsForSystems(systems []string, modesForSystems map[string][]string, trbl bootloader.TrustedAssetsBootloader, modeenv *Modeenv, includeTryModel bool) (chains []bootChain, err error) {
 	chainsForModel := func(model secboot.ModelForSealing) error {
 		modelID := modelUniqueID(model)
 		for _, system := range systems {
+			if len(modesForSystems) == 0 {
```

Comment:
I'm a bit confused why this check is not done outside of the helper?

Reply:
This was mostly causing test noise, in a lot of places we only mock some parts of the modeenv, but eg. do not set current-recover-systems or good-recovery-system in which case we'd hit this helper with an empty list of modes for systems. I'll update the code and the tests.

Reply:
Actually I've dropped this aggregate check as it's not really useful. There is a more relevant per system check further down anyway.